arch/otbn/printer.s.dfy

```diff
@@ -292,37 +293,75 @@ method printCode(c: code, depth: int)
         case Ins32(ins) => printIndent(depth); printIns32(ins);
         case Ins256(ins) => printIndent(depth); printIns256(ins);
         case Block(block) => printBlock(block, depth);
+        case Function(name, fbody) =>
+            printIndent(depth); print("jal x1, "); print(name); print("\n");
         case While(wcond, wbody) =>
         {
             printIndent(depth); printWhileCond(wcond); print(", ");
             print(codeSize(wbody) + 1); print("\n"); // + 1 for nop instruction
             printCode(wbody, depth + 1);
-            printIndent(depth+1); print("  nop\n"); // ensures different end addrs for nested loops
+            printIndent(depth+1); print("nop\n"); // ensures different end addrs for nested loops
         }
         case Comment(com) => print(com);
 }
 
-method printProc(proc_name:seq<char>, code:code)
+method getFunctionsFromCodes(block: codes, defs: set<string>, res: seq<(string, codes)>) 
+    returns (defs': set<string>, res': seq<(string, codes)>) 
 {
-    print(".globl "); print(proc_name); print("\n");
-    print(proc_name); print(":\n");
-    printCode(code, 0);
-    print("  ret\n");
+    var i := block;
+    defs', res' := defs, res;
+
+    while (i.va_CCons?)
+        decreases i
+    {
+        defs', res' := getFunctions(i.hd, defs', res');
+        i := i.tl;
+    }
 }
 
-datatype AsmTarget = OTBN
-datatype PlatformTarget = Linux | MacOS
+method getFunctions(c: code, defs: set<string>, res: seq<(string, codes)>) 
+    returns (defs': set<string>, res': seq<(string, codes)>)
+{
+    defs', res' := defs, res;
```

Comment:
Is this line redundant?

Reply:
Sort of, we can move this line into the else branch of `if name !in defs`. Otherwise it won't be recursing on the right thing.

Reply:
Ah, I see.  Yes, that seems like the more relevant place to put it.